census/estimator/trainer/model.py

```diff
@@ -0,0 +1,185 @@
+  (gender, race, education, marital_status, relationship,
+   workclass, occupation, native_country, age,
+   education_num, capital_gain, capital_loss, hours_per_week) = INPUT_COLUMNS
+  """Build an estimator."""
+  # Sparse base columns.
+  # Reused Transformations.
+  age_buckets = layers.bucketized_column(
+      age, boundaries=[18, 25, 30, 35, 40, 45, 50, 55, 60, 65])
+
+  # Wide columns and deep columns.
+  wide_columns = [
+      layers.crossed_column(
+          [education, occupation], hash_bucket_size=int(1e4)),
+      layers.crossed_column(
+          [age_buckets, race, occupation], hash_bucket_size=int(1e6)),
+      layers.crossed_column(
+          [native_country, occupation], hash_bucket_size=int(1e4)),
+      gender,
+      native_country,
+      education,
+      occupation,
+      workclass,
+      marital_status,
+      relationship,
+      age_buckets,
+  ]
+
+  deep_columns = [
+      layers.embedding_column(workclass, dimension=embedding_size),
+      layers.embedding_column(education, dimension=embedding_size),
+      layers.embedding_column(marital_status, dimension=embedding_size),
+      layers.embedding_column(gender, dimension=embedding_size),
+      layers.embedding_column(relationship, dimension=embedding_size),
+      layers.embedding_column(race, dimension=embedding_size),
+      layers.embedding_column(native_country, dimension=embedding_size),
+      layers.embedding_column(occupation, dimension=embedding_size),
+      age,
+      education_num,
+      capital_gain,
+      capital_loss,
+      hours_per_week,
+  ]
+
+  return tf.contrib.learn.DNNLinearCombinedClassifier(
+      model_dir=model_dir,
+      linear_feature_columns=wide_columns,
+      dnn_feature_columns=deep_columns,
+      dnn_hidden_units=hidden_units or [100, 70, 50, 25])
+
+
+def is_sparse(column):
+  return isinstance(column, layers.feature_column._SparseColumn)
+
+
+def feature_columns_to_placeholders(feature_columns, default_batch_size=None):
+    return {
+        column.name: tf.placeholder(
+            tf.string if is_sparse(column) else tf.float32,
+            [default_batch_size]
+        )
+        for column in feature_columns
+    }
+
+
+def json_serving_input_fn():
+    feature_placeholders = feature_columns_to_placeholders(INPUT_COLUMNS)
+    features = {
+      key: tf.expand_dims(tensor, -1)
+      for key, tensor in feature_placeholders.items()
+    }
+    return input_fn_utils.InputFnOps(
+      features,
+      None,
+      feature_placeholders
+    )
+
+
+tfrecord_serving_input_fn = tf.contrib.learn.build_parsing_serving_input_fn(
+    layers.create_feature_spec_for_parsing(INPUT_COLUMNS)
+)
+
+
+def generate_input_fn(filename,
+                      num_epochs=None,
+                      shuffle=True,
+                      skip_header_lines=0,
+                      batch_size=40):
+  def _input_fn():
+    filename_queue = tf.train.string_input_producer(
+        [filename], num_epochs=num_epochs)
```

Comment:
Irrelevant here because there's at most one file. But I'll add it anyway (and we could surface the ability to add multiple files fairly easily, which we might as well).